lib/client-assets.php

```diff
@@ -601,6 +601,56 @@ function gutenberg_register_vendor_scripts( $scripts ) {
 }
 add_action( 'wp_default_scripts', 'gutenberg_register_vendor_scripts' );
 
+/**
+ * Registers or re-registers Gutenberg Script Modules.
+ *
+ * Script modules that are registered by Core will be re-registered by Gutenberg.
+ *
+ * @since 19.3.0
+ */
+function gutenberg_default_script_modules() {
+	/*
+	 * Expects multidimensional array like:
+	 *
+	 *     'interactivity/index.min.js' => array('dependencies' => array(…), 'version' => '…'),
+	 *     'interactivity/debug.min.js' => array('dependencies' => array(…), 'version' => '…'),
+	 *     'interactivity-router/index.min.js' => …
+	 */
+	$assets = include gutenberg_dir_path() . '/build-module/assets.php';
+
+	foreach ( $assets as $file_name => $script_module_data ) {
+		/*
+		 * Build the WordPress Script Module ID from the file name.
+		 * Prepend `@wordpress/` and remove extensions and `/index` if present:
+		 *   - interactivity/index.min.js  => @wordpress/interactivity
+		 *   - interactivity/debug.min.js  => @wordpress/interactivity/debug
+		 *   - block-library/query/view.js => @wordpress/block-library/query/view
+		 */
+		$script_module_id = '@wordpress/' . preg_replace( '~(?:/index)?(?:\.min)?\.js$~iD', '', $file_name, 1 );
+		switch ( $script_module_id ) {
+			/*
+			 * Interactivity exposes two entrypoints, `/index` and `/debug`.
+			 * `/debug` should replalce `/index` in devlopment.
+			 */
+			case '@wordpress/interactivity/debug':
+				if ( ! SCRIPT_DEBUG ) {
+					continue 2;
+				}
+				$script_module_id = '@wordpress/interactivity';
+				break;
+			case '@wordpress/interactivity':
+				if ( SCRIPT_DEBUG ) {
+					continue 2;
+				}
+				break;
+		}
+
+		$path = gutenberg_url( "build-module/{$file_name}" );
```

Comment:
There is a slight change that needs to get incorporated here. There should be a different file served depending on `SCRIPT_DEBUG`:

- when `true`: `.js`
- when `false`: `.min.js`

Reply:
I'm not that sure anymore after looking at how scripts for packages are handled:

https://github.com/WordPress/gutenberg/blob/5e37a1316d0a39bc880bf6efff9b9f26db381690/lib/client-assets.php#L190-L238

Reply:
I didn't have the latest version of the branch 🙈 

This is what I see now with `npm run dev`:

<img width="1845" alt="Screenshot 2024-09-20 at 12 39 03" src="https://github.com/user-attachments/assets/4c860e70-e03d-46bf-b1cf-9bb7f1f445d4">

<img width="1729" alt="Screenshot 2024-09-20 at 12 40 06" src="https://github.com/user-attachments/assets/817932e5-1d11-4335-bc30-4142c03d568e">


I still do not think I understand it fully. `.min.js` is always produced when you use `npm run dev` and `npm run build`. `mode === 'production' && new ReadableJsAssetsWebpackPlugin(),` means that `.js` gets only generated for production builds - `npm run build`. From the docs of the webpack plugin:

> Generate a readable non-minified JS file for each `.min.js` asset.
>
> The end result is that for each JS entrypoint, we get a set of readable and non-minimized `.js` file and a minimized `.min.js`. This allows Gutenberg to follow the WordPress convention of adding a `.min.js` suffix to minimized JS files, while still providing a readable and unminized files that play well with the WordPress i18n machinery.

I'm still puzzled why there is no `SCRIPT_DEBUG` involved like in WP core.

Reply:
I think we can remove [44da0b0](https://github.com/WordPress/gutenberg/pull/65460/commits/44da0b0caee077fdc5318ebaa8a66eb33c42847b) for now and ship it all as is.

Reply:
The same folder after `npm run build`:

<img width="1783" alt="Screenshot 2024-09-20 at 12 49 23" src="https://github.com/user-attachments/assets/9645faa6-b8c8-47a5-b144-54b71d0088ef">

Notice `.js` and `.min.js` files get created. For `npm run `dev`, I see only `.min.js` files 🤷🏻 In effect, it behaves like in debug mode. However, once the plugin gets released, there is no longer a way to use the debug version. In fact, it probably doesn't make too much sense, because `.js` file is built with production mode.

Reply:
Yes, it's confusing. When doing a development build, `.min.js` files are produced but they're not minified. The SCRIPT_DEBUG doesn't change the files Gutenberg uses.

I've reverted the `.min.js`/`.js` changes.